File: backend/services/feedbackService.js
const { createRailwayClient } = require('../config/supabase');
const logger = require('../utils/logger');

class FeedbackService {
    constructor() {
        this.table = 'customer_feedback';
        this.templatesTable = 'message_templates';
        this.responsesTable = 'feedback_responses';
        this.chatsTable = 'active_chats';
        this.messagesTable = 'chat_messages';
        this.campaignsTable = 'mass_campaigns';
    }

    // Categorização automática de feedback
    async categorizeFeedback(message) {
        try {
            // Análise simples de sentimentos baseada em palavras-chave
            const positiveWords = ['ótimo', 'excelente', 'bom', 'maravilhoso', 'perfeito', 'satisfeito', 'feliz', 'bom serviço', 'recomendo', 'parabéns'];
            const negativeWords = ['ruim', 'péssimo', 'horrível', 'terrível', 'insatisfeito', 'problema', 'reclamação', 'demora', 'erro', 'defeito', 'queixa'];
            
            const lowerMessage = message.toLowerCase();
            let positiveScore = 0;
            let negativeScore = 0;
            
            positiveWords.forEach(word => {
                if (lowerMessage.includes(word)) positiveScore++;
            });
            
            negativeWords.forEach(word => {
                if (lowerMessage.includes(word)) negativeScore++;
            });
            
            let category = 'neutral';
            let sentimentScore = 0.0;
            
            if (positiveScore > negativeScore) {
                category = 'positive';
                sentimentScore = Math.min(positiveScore / 3, 1.0);
            } else if (negativeScore > positiveWords) {
                category = 'negative';
                sentimentScore = -Math.min(negativeScore / 3, 1.0);
            }
            
            return { category, sentimentScore };
        } catch (error) {
            logger.error('Erro na categorização de feedback:', error);
            return { category: 'neutral', sentimentScore: 0.0 };
        }
    }

    // Adicionar novo feedback
    async addFeedback(clientId, message, source = 'whatsapp', isManual = false) {
        try {
            const { category, sentimentScore } = await this.categorizeFeedback(message);
            
            const { data, error } = await supabase
                .from(this.table)
                .insert({
                    client_id: clientId,
                    message,
                    category,
                    sentiment_score: sentimentScore,
                    source,
                    is_manual: isManual
                })
                .select()
                .single();

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao adicionar feedback:', error);
            throw error;
        }
    }

    // Buscar feedbacks com filtros
    async getFeedback(filters = {}) {
        try {
            let query = supabase
                .from(this.table)
                .select(`*, clients(client_name, phone, additional_info)`)
                .order('created_at', { ascending: false });

            if (filters.category) {
                query = query.eq('category', filters.category);
            }

            if (filters.clientId) {
                query = query.eq('client_id', filters.clientId);
            }

            if (filters.dateFrom) {
                query = query.gte('created_at', filters.dateFrom);
            }

            if (filters.dateTo) {
                query = query.lte('created_at', filters.dateTo);
            }

            const { data, error } = await query;
            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao buscar feedback:', error);
            throw error;
        }
    }

    // Atualizar categoria do feedback
    async updateFeedbackCategory(feedbackId, category, isManual = true) {
        try {
            const { data, error } = await supabase
                .from(this.table)
                .update({ category, is_manual: isManual })
                .eq('id', feedbackId)
                .select()
                .single();

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao atualizar categoria:', error);
            throw error;
        }
    }

    // Templates de mensagens
    async getMessageTemplates(category = null) {
        try {
            // Sempre usar templates padrão pois Supabase foi removido
            console.log('Usando templates padrão (Supabase não configurado)');
            return this.getDefaultTemplates();
        } catch (error) {
            logger.error('Erro ao buscar templates:', error.message);
            return this.getDefaultTemplates();
        }
    }

    // Templates padrão quando Supabase não está disponível
    getDefaultTemplates() {
        return [
            {
                id: 'default_thanks',
                name: 'Agradecimento Simples',
                category: 'positive',
                template: 'Obrigado pelo seu feedback, {{client_name}}!',
                variables: ['client_name']
            },
            {
                id: 'default_followup',
                name: 'Acompanhamento',
                category: 'neutral',
                template: 'Olá {{client_name}}, gostaríamos de entender melhor seu feedback.',
                variables: ['client_name']
            },
            {
                id: 'default_apology',
                name: 'Desculpas',
                category: 'negative',
                template: 'Lamentamos {{client_name}}, vamos melhorar!',
                variables: ['client_name']
            }
        ];
    }

    async createMessageTemplate(name, category, template, variables = []) {
        try {
            const { data, error } = await supabase
                .from(this.templatesTable)
                .insert({ name, category, template, variables })
                .select()
                .single();

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao criar template:', error);
            throw error;
        }
    }

    // Gerar resposta baseada em template
    async generateResponse(clientId, templateId, feedbackMessage, customData = {}) {
        try {
            if (!this.supabase) {
                throw new Error('Supabase não configurado - não é possível gerar resposta');
            }

            const [template, client] = await Promise.all([
                this.supabase.from(this.templatesTable).select('*').eq('id', templateId).single(),
                this.supabase.from('clients').select('*').eq('id', clientId).single()
            ]);

            if (template.error) throw template.error;
            if (client.error) throw client.error;

            let response = template.data.template;
            const data = {
                client_name: client.data.client_name,
                ...customData
            };

            // Substituir variáveis no template
            Object.keys(data).forEach(key => {
                response = response.replace(new RegExp(`{{${key}}}`, 'g'), data[key] || '');
            });

            return response;
        } catch (error) {
            logger.error('Erro ao gerar resposta:', error);
            throw error;
        }
    }

    // Sistema de chat
    async startChat(clientId) {
        try {
            const { data, error } = await supabase
                .from(this.chatsTable)
                .insert({ client_id: clientId })
                .select()
                .single();

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao iniciar chat:', error);
            throw error;
        }
    }

    async getChatHistory(clientId) {
        try {
            const { data, error } = await supabase
                .from(this.chatsTable)
                .select(`*, chat_messages(*), clients(*)`)
                .eq('client_id', clientId)
                .order('started_at', { ascending: false });

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao buscar histórico:', error);
            throw error;
        }
    }

    async sendChatMessage(chatId, sender, message, messageType = 'text') {
        try {
            const { data, error } = await supabase
                .from(this.messagesTable)
                .insert({
                    chat_id: chatId,
                    sender,
                    message,
                    message_type: messageType
                })
                .select()
                .single();

            if (error) throw error;

            // Atualizar última mensagem do chat
            await supabase
                .from(this.chatsTable)
                .update({ last_message_at: new Date().toISOString() })
                .eq('id', chatId);

            return data;
        } catch (error) {
            logger.error('Erro ao enviar mensagem:', error);
            throw error;
        }
    }

    // Campanhas de mensagens em massa
    async createMassCampaign(name, templateId, messageContent, targetFilter = {}) {
        try {
            const { data, error } = await supabase
                .from(this.campaignsTable)
                .insert({
                    name,
                    template_id: templateId,
                    message_content: messageContent,
                    target_filter: targetFilter
                })
                .select()
                .single();

            if (error) throw error;
            return data;
        } catch (error) {
            logger.error('Erro ao criar campanha:', error);
            throw error;
        }
    }

    async getCampaignTargets(campaignId) {
        try {
            if (!this.supabase) {
                throw new Error('Supabase não configurado - não é possível buscar alvos da campanha');
            }

            const campaign = await this.supabase
                .from(this.campaignsTable)
                .select('*')
                .eq('id', campaignId)
                .single();

            if (campaign.error) throw campaign.error;

            const filter = campaign.data.target_filter;
            let query = this.supabase.from('clients').select('*');

            // Aplicar filtros
            if (filter.region) {
                query = query.ilike('additional_info->>region', `%${filter.region}%`);
            }

            if (filter.dateFrom) {
                query = query.gte('created_at', filter.dateFrom);
            }

            if (filter.dateTo) {
                query = query.lte('created_at', filter.dateTo);
            }

            const { data, error } = await query;
            if (error) throw error;

            return data;
        } catch (error) {
            logger.error('Erro ao buscar alvos:', error);
            throw error;
        }
    }

    async executeMassCampaign(campaignId) {
        try {
            const targets = await this.getCampaignTargets(campaignId);
            const campaign = await supabase
                .from(this.campaignsTable)
                .select('*')
                .eq('id', campaignId)
                .single();

            if (campaign.error) throw campaign.error;

            let sentCount = 0;
            let failedCount = 0;

            for (const client of targets) {
                try {
                    const personalizedMessage = campaign.data.message_content
                        .replace(/{{client_name}}/g, client.client_name)
                        .replace(/{{phone}}/g, client.phone);

                    // Aqui você integraria com o serviço WhatsApp
                    // Por enquanto, apenas registramos como enviado
                    await supabase
                        .from('campaign_contacts')
                        .insert({
                            campaign_id: campaignId,
                            client_id: client.id,
                            status: 'sent',
                            sent_at: new Date().toISOString()
                        });

                    sentCount++;
                } catch (error) {
                    await supabase
                        .from('campaign_contacts')
                        .insert({
                            campaign_id: campaignId,
                            client_id: client.id,
                            status: 'failed',
                            error_message: error.message
                        });
                    failedCount++;
                }
            }

            await supabase
                .from(this.campaignsTable)
                .update({
                    status: 'completed',
                    sent_contacts: sentCount,
                    failed_contacts: failedCount,
                    completed_at: new Date().toISOString()
                })
                .eq('id', campaignId);

            return { success: true, sentCount, failedCount };
        } catch (error) {
            logger.error('Erro ao executar campanha:', error);
            throw error;
        }
    }

    // Estatísticas de feedback
    async getFeedbackStats() {
        try {
            // Retornar estatísticas padrão quando Supabase não estiver configurado
            if (!this.supabase) {
                return {
                    total: 0,
                    byCategory: {},
                    recent: []
                };
            }

            const [total, allFeedback, recent] = await Promise.all([
                this.supabase.from(this.table).select('count', { count: 'exact' }),
                this.supabase.from(this.table).select('category'),
                this.supabase.from(this.table).select('*').limit(10).order('created_at', { ascending: false })
            ]);

            // Agrupar manualmente por categoria
            const categoryCounts = {};
            if (allFeedback.data) {
                allFeedback.data.forEach(item => {
                    const category = item.category || 'uncategorized';
                    categoryCounts[category] = (categoryCounts[category] || 0) + 1;
                });
            }

            const byCategory = Object.entries(categoryCounts).map(([category, count]) => ({
                category,
                count
            }));

            return {
                total: total.data?.[0]?.count || 0,
                byCategory,
                recent: recent.data || []
            };
        } catch (error) {
            logger.error('Erro ao buscar estatísticas:', error);
            throw error;
        }
    }

    // Salvar apenas respostas aos feedbacks (não salva mensagens de chat)
    async saveFeedbackResponse({ contactId, contactName, response, timestamp, type }) {
        try {
            const { data, error } = await supabase
                .from('feedback_responses')
                .insert({
                    contact_id: contactId,
                    contact_name: contactName,
                    response: response,
                    response_type: type || 'whatsapp_response',
                    created_at: timestamp || new Date().toISOString()
                })
                .select()
                .single();

            if (error) throw error;
            
            logger.info('Resposta ao feedback salva:', { contactId, contactName });
            return data;
        } catch (error) {
            logger.error('Erro ao salvar resposta ao feedback:', error);
            
            // Fallback: salvar em arquivo local se Supabase não estiver disponível
            const fs = require('fs');
            const path = require('path');
            const responsesFile = path.join(__dirname, '..', 'data', 'feedback_responses.json');
            
            // Criar diretório se não existir
            const dataDir = path.dirname(responsesFile);
            if (!fs.existsSync(dataDir)) {
                fs.mkdirSync(dataDir, { recursive: true });
            }
            
            // Ler arquivo existente ou criar novo
            let responses = [];
            if (fs.existsSync(responsesFile)) {
                try {
                    responses = JSON.parse(fs.readFileSync(responsesFile, 'utf8'));
                } catch (e) {
                    responses = [];
                }
            }
            
            // Adicionar nova resposta
            const newResponse = {
                id: Date.now(),
                contact_id: contactId,
                contact_name: contactName,
                response: response,
                response_type: type || 'whatsapp_response',
                created_at: timestamp || new Date().toISOString()
            };
            
            responses.push(newResponse);
            fs.writeFileSync(responsesFile, JSON.stringify(responses, null, 2));
            
            return newResponse;
        }
    }
}

module.exports = new FeedbackService();

File: backend/app/services/feedback_service.py
#!/usr/bin/env python3
"""
Serviço de Feedback - Gerencia feedbacks no banco de dados
"""

import json
from datetime import datetime
from typing import List, Dict, Optional
import logging
from .sentiment_analyzer import sentiment_analyzer

logger = logging.getLogger(__name__)

class FeedbackService:
    def __init__(self, db_manager=None):
        self.db_manager = db_manager
        self.table_name = 'feedbacks'
        
    def create_feedback_table(self):
        """Cria a tabela de feedbacks se não existir"""
        try:
            if not self.db_manager:
                logger.warning("DB Manager não disponível, usando fallback")
                return False
                
            query = """
            CREATE TABLE IF NOT EXISTS feedbacks (
                id SERIAL PRIMARY KEY,
                feedback_id VARCHAR(50) UNIQUE,
                contact_name VARCHAR(100),
                contact_phone VARCHAR(20),
                text TEXT,
                sentiment VARCHAR(20),
                score DECIMAL(5,4),
                keywords JSON,
                timestamp TIMESTAMP,
                analyzed_at TIMESTAMP,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
            """
            
            self.db_manager.execute_query(query)
            logger.info("Tabela de feedbacks criada/verificada")
            return True
            
        except Exception as e:
            logger.error(f"Erro ao criar tabela de feedbacks: {e}")
            return False

    def save_feedback(self, feedback_data: Dict) -> bool:
        """Salva um feedback no banco de dados"""
        try:
            if not self.db_manager:
                logger.warning("DB Manager não disponível, usando fallback")
                return False
                
            # Garante que a tabela existe
            self.create_feedback_table()
            
            query = """
            INSERT INTO feedbacks (
                feedback_id, contact_name, contact_phone, text, 
                sentiment, score, keywords, timestamp, analyzed_at
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (feedback_id) DO UPDATE SET
                contact_name = EXCLUDED.contact_name,
                contact_phone = EXCLUDED.contact_phone,
                text = EXCLUDED.text,
                sentiment = EXCLUDED.sentiment,
                score = EXCLUDED.score,
                keywords = EXCLUDED.keywords,
                timestamp = EXCLUDED.timestamp,
                analyzed_at = EXCLUDED.analyzed_at
            """
            
            values = (
                feedback_data.get('id'),
                feedback_data.get('contact_name'),
                feedback_data.get('contact_phone'),
                feedback_data.get('text'),
                feedback_data.get('sentiment'),
                feedback_data.get('score'),
                json.dumps(feedback_data.get('keywords', [])),
                feedback_data.get('timestamp'),
                feedback_data.get('analyzed_at')
            )
            
            self.db_manager.execute_query(query, values)
            logger.info(f"Feedback salvo: {feedback_data.get('id')}")
            return True
            
        except Exception as e:
            logger.error(f"Erro ao salvar feedback: {e}")
            return False

    def get_all_feedbacks(self, limit: int = 100) -> List[Dict]:
        """Busca todos os feedbacks do banco"""
        try:
            if not self.db_manager:
                logger.warning("DB Manager não disponível, retornando lista vazia")
                return []
                
            query = """
            SELECT 
                feedback_id, contact_name, contact_phone, text,
                sentiment, score, keywords, timestamp, analyzed_at
            FROM feedbacks 
            ORDER BY timestamp DESC 
            LIMIT %s
            """
            
            results = self.db_manager.fetch_all(query, (limit,))
            
            feedbacks = []
            for row in results:
                feedback = {
                    'id': row[0],
                    'contact_name': row[1],
                    'contact_phone': row[2],
                    'text': row[3],
                    'sentiment': row[4],
                    'score': float(row[5]) if row[5] else 0.0,
                    'keywords': json.loads(row[6]) if row[6] else [],
                    'timestamp': row[7].isoformat() if row[7] else None,
                    'analyzed_at': row[8].isoformat() if row[8] else None
                }
                feedbacks.append(feedback)
            
            logger.info(f"Retornados {len(feedbacks)} feedbacks")
            return feedbacks
            
        except Exception as e:
            logger.error(f"Erro ao buscar feedbacks: {e}")
            return []

    def get_feedbacks_by_sentiment(self, sentiment: str, limit: int = 50, days: int = None) -> List[Dict]:
        """Busca feedbacks por sentimento e período"""
        try:
            if not self.db_manager:
                return []
            
            if days:
                query = """
                SELECT 
                    feedback_id, contact_name, contact_phone, text,
                    sentiment, score, keywords, timestamp, analyzed_at
                FROM feedbacks 
                WHERE sentiment = %s AND timestamp >= NOW() - INTERVAL '%s days'
                ORDER BY timestamp DESC 
                LIMIT %s
                """
                results = self.db_manager.fetch_all(query, (sentiment, days, limit))
            else:
                query = """
                SELECT 
                    feedback_id, contact_name, contact_phone, text,
                    sentiment, score, keywords, timestamp, analyzed_at
                FROM feedbacks 
                WHERE sentiment = %s
                ORDER BY timestamp DESC 
                LIMIT %s
                """
                results = self.db_manager.fetch_all(query, (sentiment, limit))
            
            feedbacks = []
            for row in results:
                feedback = {
                    'id': row[0],
                    'contact_name': row[1],
                    'contact_phone': row[2],
                    'text': row[3],
                    'sentiment': row[4],
                    'score': float(row[5]) if row[5] else 0.0,
                    'keywords': json.loads(row[6]) if row[6] else [],
                    'timestamp': row[7].isoformat() if row[7] else None,
                    'analyzed_at': row[8].isoformat() if row[8] else None
                }
                feedbacks.append(feedback)
            
            return feedbacks
            
        except Exception as e:
            logger.error(f"Erro ao buscar feedbacks por sentimento: {e}")
            return []

    def get_feedbacks_by_date_range(self, days: int, limit: int = 100) -> List[Dict]:
        """Busca feedbacks por período de dias"""
        try:
            if not self.db_manager:
                return []
                
            query = """
            SELECT 
                feedback_id, contact_name, contact_phone, text,
                sentiment, score, keywords, timestamp, analyzed_at
            FROM feedbacks 
            WHERE timestamp >= NOW() - INTERVAL '%s days'
            ORDER BY timestamp DESC 
            LIMIT %s
            """
            
            results = self.db_manager.fetch_all(query, (days, limit))
            
            feedbacks = []
            for row in results:
                feedback = {
                    'id': row[0],
                    'contact_name': row[1],
                    'contact_phone': row[2],
                    'text': row[3],
                    'sentiment': row[4],
                    'score': float(row[5]) if row[5] else 0.0,
                    'keywords': json.loads(row[6]) if row[6] else [],
                    'timestamp': row[7].isoformat() if row[7] else None,
                    'analyzed_at': row[8].isoformat() if row[8] else None
                }
                feedbacks.append(feedback)
            
            return feedbacks
            
        except Exception as e:
            logger.error(f"Erro ao buscar feedbacks por período: {e}")
            return []

    def get_feedback_stats_by_date(self, days: int = None) -> Dict:
        """Retorna estatísticas dos feedbacks por período"""
        try:
            if not self.db_manager:
                return {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}
            
            if days:
                query = """
                SELECT 
                    sentiment,
                    COUNT(*) as count
                FROM feedbacks 
                WHERE timestamp >= NOW() - INTERVAL '%s days'
                GROUP BY sentiment
                """
                results = self.db_manager.fetch_all(query, (days,))
            else:
                query = """
                SELECT 
                    sentiment,
                    COUNT(*) as count
                FROM feedbacks 
                GROUP BY sentiment
                """
                results = self.db_manager.fetch_all(query)
            
            stats = {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}
            
            for row in results:
                sentiment = row[0]
                count = row[1]
                stats[sentiment] = count
                stats['total'] += count
            
            return stats
            
        except Exception as e:
            logger.error(f"Erro ao buscar estatísticas por período: {e}")
            return {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}

    def get_feedback_stats(self) -> Dict:
        """Retorna estatísticas dos feedbacks"""
        try:
            if not self.db_manager:
                return {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}
                
            query = """
            SELECT 
                sentiment,
                COUNT(*) as count
            FROM feedbacks 
            GROUP BY sentiment
            """
            
            results = self.db_manager.fetch_all(query)
            
            stats = {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}
            
            for row in results:
                sentiment = row[0]
                count = row[1]
                stats[sentiment] = count
                stats['total'] += count
            
            return stats
            
        except Exception as e:
            logger.error(f"Erro ao buscar estatísticas: {e}")
            return {'positive': 0, 'negative': 0, 'neutral': 0, 'total': 0}

    def analyze_and_save_message(self, message_data: Dict) -> Dict:
        """Analisa uma mensagem e salva como feedback"""
        try:
            # Analisa a mensagem
            feedback_data = sentiment_analyzer.analyze_message(message_data)
            
            if not feedback_data:
                logger.error("Falha na análise da mensagem")
                return {}
            
            # Salva no banco
            success = self.save_feedback(feedback_data)
            
            if success:
                logger.info(f"Feedback analisado e salvo: {feedback_data.get('id')}")
                return feedback_data
            else:
                logger.error("Falha ao salvar feedback")
                return feedback_data  # Retorna mesmo sem salvar
                
        except Exception as e:
            logger.error(f"Erro ao analisar e salvar mensagem: {e}")
            return {}

# Instância global do serviço
feedback_service = FeedbackService()

